prov/efa/src/rdm/efa_rdm_rma.c

```diff
@@ -173,8 +173,10 @@ ssize_t efa_rdm_rma_readmsg(struct fid_ep *ep, const struct fi_msg_rma *msg, uin
 
 	/*
 	 * A handshake is required to choose the correct protocol (whether to use device read).
+	 * For local read (read from self ep), such handshake is not needed because we only
+	 * need to check the local ep's capabilities.
 	 */
-	if (!(peer->flags & EFA_RDM_PEER_HANDSHAKE_RECEIVED)) {
+	if (!(peer->is_self) && !(peer->flags & EFA_RDM_PEER_HANDSHAKE_RECEIVED)) {
 		err = efa_rdm_ep_trigger_handshake(efa_rdm_ep, txe->addr);
```

Comment:
Do you think `efa_rdm_ep_trigger_handshake()` itself should prevent handshakes when `peer->is_self` is true?

Reply:
Don't think so. this function just shouldn't be called when peer is self. We should make sure `peer->is_self` is false inside the function, that can be an `assert()`

Reply:
The special thing for this read and write op is we enforce a handshake and return EAGAIN before the handshake is made, which can cause extra delay. In that regard preventing self peer to call this function and return EAGAIN is essential